src/Microsoft.Diagnostics.ExtensionCommands/ClrMDHelper.cs

```diff
@@ -733,7 +733,7 @@ private bool TryGetSegmentMemoryRange(ClrSegment segment, GCGeneration generatio
                     }
                     return start != end;
                 case GCGeneration.Generation2:
-                    if (!segment.IsLargeObjectSegment)
+                    if (!(segment.IsLargeObjectSegment || segment.IsPinnedObjectSegment))
```

Comment:
This behavior is a bit strange to me.

For repro:
```c#
using System;

var buffer = GC.AllocateUninitializedArray<byte>(10_000, pinned: true);

Console.WriteLine("Dump now");
Console.ReadKey();

Console.WriteLine(buffer.Length);
```
| Generation | Status |
| --- | --- |
| Gen 0 | lots of object |
| Gen 1 | should be empty |
| Gen 2 | should be empty |
| LOH | should be empty |
| POH | `00007ffad1bbb718        1        10024 System.Byte[]` and some static roots (?) |

Without that change Gen 2 shows the contents of POH too, which isn't correct.

TBH I'd expect `ClrSegment.Generation2` to get the correct memroy range, which (to my knowledge) should be distinct from LOH and POH. So the whole check for `if (!segment.IsLargeObjectSegment)` (current main) or `if (!(segment.IsLargeObjectSegment || segment.IsPinnedObjectSegment))` (PR) sounds like a workaround to me.
Should this be investigated in ClrMD instead?

Reply:
The `object[]` you are seeing for static root is probably due to https://github.com/dotnet/runtime/pull/47651

Reply:
> The `object[]` you are seeing for static root is probably due to [dotnet/runtime#47651](https://github.com/dotnet/runtime/pull/47651)

That was / is my guess.

The point that concerns me is that `ClrSegment.Generation2` from ClrMD shows objects that are in the POH. POH is known to ClrMD.
Similar why must IsLargeObjectSegment be explicitely checked, if it is a distinct segment on the GC-heap?

Reply:
this is explained in [mem-doc](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#loh-large-object-heap) -

> The LOH is internally kept tracked as gen3 but logically it's part of gen2, meaning that the LOH is collected only during gen2 GCs. This means if you churn the LOH a lot you'd be triggering gen2 GCs very often and if your gen2 is also big that means GC will have to do a lot of work performing gen2 GCs.

I will add to the doc that POH has the same characteristics.

Reply:
Thanks!

LOH as well as POH are distinct segements from Gen2?
If so ClrMD should handle it that way. Do I have a misconception here?

Reply:
yes, LOH occupies separate segs from gen2 and so does POH.

Reply:
ClrMD has always called objects on the LoH "gen2" since they are logically part of gen2.  When the Pinned Object Heap was introduced we similarly just called all of the objects there gen2 objects.  You always have the option to check ClrSegment's `IsLargeObjectSegment` and `IsPinnedObjectSegment` if you need an escape valve to know that "hey these aren't _real_ gen2 objects" and since the LoH and PoH will be entirely filled as `Generation2` you can call those objects anything you like.

Reply:
yes, the ClrMd way of handling this is how it's supposed to be handled. @gfoidl did you mean you wanted it to be handled some other way? I can't tell from your comment.

Reply:
> did you mean you wanted it to be handled some other way?

My understanding (before this conversation) was that ClrSegment.Generation2 only has objects that are in the segs of Gen2, and LOH's and POH's segs are different one that the ones from Gen2.
Thus there should be ClrSegment.LargeObjectHeap and ClrSegment.PinnedObjectHeap, as the segs are distinct.

I knew that logically LOH and POH belong to Gen2 ("collected together"), but didn't think about that the segments also count as Gen2.

Now I see ClrMD does the right thing. Gen2 and to have further distinctino there are the properties for IsLOH and IsPOH.